### PR TITLE
MINIFICPP-2047 added reverseDnsLookup to Expression Language

### PR DESCRIPTION
**Pull request by martinzink**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by adamdebreceni — APPROVED** [review]

File: EXPRESSIONS.md [hunk 1]
```diff
@@ -1590,6 +1591,31 @@ to evaluate additional functions against.
 ${allMatchingAttributes('a.*'):count()} ):gt(3)}` returns true if there are
 more than 3 attributes whose names begin with the letter a.
 
+### reverseLookup
```
Comment (fgerlits) [inline EXPRESSIONS.md:1594]: maybe we could call this `reverseDNSLookup`?  `reverseLookup` sounds more general than it really is

Comment (martinzink) [reply EXPRESSIONS.md]: good point, I've renamed it in https://github.com/apache/nifi-minifi-cpp/pull/1510/commits/31dacdb37bcf4cdad0f32d6311a047f549920dd4

**Review by lordgamez — APPROVED** [review]

**Review by fgerlits — APPROVED** [review]

File: extensions/expression-language/Expression.cpp [hunk 2]
```diff
@@ -183,6 +187,20 @@ Value expr_ip(const std::vector<Value>& /*args*/) {
...
+
+  std::chrono::steady_clock::duration timeout_duration = 5s;
+  if (args.size() > 1) {
+    timeout_duration = std::chrono::milliseconds(args[1].asUnsignedLong());
+  }
+
+  return utils::net::addressFromString(ip_address_str)
+      | utils::flatMap([timeout_duration](const auto& ip_address) { return utils::net::reverseDnsLookup(ip_address, timeout_duration);})
+      | utils::map([](const auto& hostname)-> Value { return Value(hostname); })
+      | utils::valueOrElse([&](std::error_code error_code) { throw std::runtime_error(error_code.message());});
```
Comment (szaszm) [inline extensions/expression-language/Expression.cpp:201]: I'd throw a `std::system_error` instead. It's meant to wrap an error code, as opposed to the more general `std::runtime_error`.

File: libminifi/include/utils/net/DNS.h [hunk 3]
```diff
@@ -15,13 +15,16 @@
  * limitations under the License.
  */
 #pragma once
+#include <chrono>
 #include <memory>
 #include <string>
 #include <string_view>
 #include <system_error>
+#include <vector>
```
Comment (szaszm) [inline libminifi/include/utils/net/DNS.h:23]: I don't see the need to include vector.

Comment (martinzink) [reply libminifi/include/utils/net/DNS.h]: You are right, removed it in https://github.com/apache/nifi-minifi-cpp/pull/1510/commits/4bb1c8fc0d9c456613ef49a2bdc33762b2caf3d5#diff-f0907daecdf93b300b66d2224cf21f677c966e82a58b738132721daf27852c26L23

File: extensions/expression-language/Expression.cpp (hunk 2 above)
Comment (martinzink) [reply extensions/expression-language/Expression.cpp]: Good idea, I've changed it in https://github.com/apache/nifi-minifi-cpp/pull/1510/commits/4bb1c8fc0d9c456613ef49a2bdc33762b2caf3d5#diff-f84ab4bed4b25e41f8cb612c6444720295b6050fdb70720ada4aea04b389565aR201

**Pull request closed without merging**